commit 2337d8a4f69c3603fc214416a069c6eace7f245a (refs/changes/17/4917/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2018-10-02T17:29:37-07:00 (1 year ago)
    
    TRITON-825 cns-server not forgiving enough of slow updates on replicas

diff --git a/lib/dns-server.js b/lib/dns-server.js
index ce099cf..1798905 100644
--- a/lib/dns-server.js
+++ b/lib/dns-server.js
@@ -59,6 +59,7 @@ function DNSServer(opts) {
 		this.notifyInterval = consts.DEFAULT_MIN_NOTIFY_INT;
 	this.checkAndNotify();
 	this.unblacklistTimers = {};
+	this.checkTimers = {};
 
 	this.peerCounters = {};
 
@@ -1072,20 +1073,33 @@ DNSServer.prototype.handleTransfer = function (q, z, cb) {
 			q.addAnswer(z, soa, TTL);
 
 			var log  = q.log;
-			setTimeout(function () {
+			if (addr !== '127.0.0.1' && addr !== '::1')
+				ensurePeerCheckTimer();
+			q.send();
+			cb();
+			q.log.info('responded ok');
+
+			function ensurePeerCheckTimer() {
+				if (!self.checkTimers[normalizeIP(addr)]) {
+					self.checkTimers[normalizeIP(addr)] =
+					    setTimeout(doPeerChecks, 2000);
+				}
+			}
+
+			function doPeerChecks() {
 				var arg = {};
 				vasync.pipeline({
 					funcs: [claim, checkSOA, checkVer],
 					arg: arg
 				}, function () {
+					delete (self.checkTimers[
+					    normalizeIP(addr)]);
+					if (arg.rearm)
+						ensurePeerCheckTimer();
 					if (arg.handle)
 						arg.handle.release();
 				});
-			}, 1000);
-
-			q.send();
-			cb();
-			q.log.info('responded ok');
+			}
 
 			function claim(_, ccb) {
 				self.redisPool.claim(
@@ -1113,7 +1127,10 @@ DNSServer.prototype.handleTransfer = function (q, z, cb) {
 						    'xfer follow-up showed ' +
 						    'peer still stuck on ' +
 						    'serial %d', qsoa.serial);
-						ccb(new Error('Stuck'));
+						_.rearm = true;
+						_.redis.hset(
+						    'peer:' + normalizeIP(addr),
+						    z, String(qsoa.serial), ccb);
 						return;
 					}
 					self.incrPeerCounter(addr, 'goodxfer');
